app/models/concerns/profile_tailoring.rb

```diff
@@ -33,4 +33,20 @@ def update_os_minor_version(version)
 
     update!(os_minor_version: version)
   end
+
+  def tailoring_valid?
+    rule_groups = RuleGroup.where(id: rules.includes(:rule_group)
+                           .references(:rule_group).flat_map { |r| r.rule_group&.path_ids })
+
+    conflicts, requires = %w[conflicts requires].map do |relationship|
+      with_relationships(relationship, rules, rule_groups).map(&:right)
+    end
+
+    valid_relationships?(rules, rule_groups, requires, conflicts)
```

Comment:
As I said earlier, extracting the end of a method chain into an external method is a bit complicated. You want a dataflow in the opposite direction, i.e. extract methods from the beginning and make the end the simplest as possible by just aggregating it there. Rubocop is just giving you hints and it's not always the best to start cleaning up a method from the end. So please try to extract the `rule_groups =` into `RulesAndRuleGroups#hierarchical_rule_groups` or `rule_groups_with_children`. I think you don't need to memoize it in a variable, as the SQL cache kicks in anyway. Then this `valid_relationships?` can be moved back here as line 38 and 39 will completely go away.

Reply:
Ahh, I misunderstood that previous comment about the `rule_groups =` part but I see what you mean now. I'll update this.